File: app01_java_block/src/test/java/com/devpowers/legalchain/core/BlockchainTest.java
package com.devpowers.legalchain.core;

import com.devpowers.legalchain.consensus.BftStrategy;
import com.devpowers.legalchain.consensus.ConsensusEngine;
import com.devpowers.legalchain.consensus.ProofOfStakeStrategy;
import com.devpowers.legalchain.consensus.ProofOfWorkStrategy;
import com.devpowers.legalchain.crypto.PqcSignatureService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.security.KeyPair;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

/**
 * Proves the "real ledger, not a mock" properties: signed submission, sealing with
 * rewards, full-chain audit, tamper detection and the longest-valid-chain sync rule.
 */
class BlockchainTest {

    private final PqcSignatureService signatures = new PqcSignatureService();
    private Blockchain blockchain;
    private KeyPair sender;
    private String senderAddress;

    @BeforeEach
    void setUp() {
        ConsensusEngine engine = new ConsensusEngine(
                List.of(new ProofOfWorkStrategy(), new ProofOfStakeStrategy(), new BftStrategy()),
                "POS");
        TokenomicsService tokenomics = new TokenomicsService(
                new BigDecimal("50"), 100, new BigDecimal("21000"));
        blockchain = new Blockchain(engine, tokenomics, signatures, event -> {
        });
        sender = signatures.generateKeyPair();
        senderAddress = signatures.fingerprint(sender.getPublic());
    }

    private Transaction signedTransfer(String recipient, String amount) {
        Transaction unsigned = new Transaction(
                UUID.randomUUID().toString(), System.currentTimeMillis(),
                senderAddress, recipient, new BigDecimal(amount),
                TransactionType.TRANSFER, Map.of(),
                signatures.encodePublicKey(sender.getPublic()), null);
        return new Transaction(
                unsigned.id(), unsigned.timestamp(), unsigned.sender(), unsigned.recipient(),
                unsigned.amount(), unsigned.type(), unsigned.payload(),
                unsigned.senderPublicKey(), signatures.sign(unsigned.contentToSign(), sender.getPrivate()));
    }

    @Test
    void startsWithDeterministicGenesis() {
        assertThat(blockchain.length()).isEqualTo(1);
        assertThat(blockchain.blocks().getFirst().previousHash()).isEqualTo("0".repeat(64));
        assertThat(blockchain.isValid()).isTrue();
    }

    @Test
    void rejectsUnsignedAndForgedTransactions() {
        Transaction unsigned = new Transaction(
                UUID.randomUUID().toString(), System.currentTimeMillis(),
                senderAddress, "bob", BigDecimal.ONE,
                TransactionType.TRANSFER, Map.of(), null, null);
        assertThatThrownBy(() -> blockchain.submit(unsigned))
                .isInstanceOf(IllegalArgumentException.class);

        // Valid signature but claimed sender is not the signing key's fingerprint.
        Transaction impersonation = new Transaction(
                UUID.randomUUID().toString(), System.currentTimeMillis(),
                "someone-else", "bob", BigDecimal.ONE,
                TransactionType.TRANSFER, Map.of(),
                signatures.encodePublicKey(sender.getPublic()), null);
        Transaction signedImpersonation = new Transaction(
                impersonation.id(), impersonation.timestamp(), impersonation.sender(),
                impersonation.recipient(), impersonation.amount(), impersonation.type(),
                impersonation.payload(), impersonation.senderPublicKey(),
                signatures.sign(impersonation.contentToSign(), sender.getPrivate()));
        assertThatThrownBy(() -> blockchain.submit(signedImpersonation))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void minesBlockWithRewardAndDerivesBalances() {
        blockchain.submit(signedTransfer("bob", "3"));
        Block block = blockchain.minePendingTransactions("validator-x");

        assertThat(block.index()).isEqualTo(1);
        assertThat(block.transactions()).hasSize(2); // transfer + reward
        assertThat(blockchain.isValid()).isTrue();
        assertThat(blockchain.pendingTransactions()).isEmpty();
        assertThat(blockchain.balances())
                .containsEntry("bob", new BigDecimal("3"))
                .containsEntry("validator-x", new BigDecimal("50"));
    }

    @Test
    void detectsTamperedHistory() {
        blockchain.submit(signedTransfer("bob", "3"));
        blockchain.minePendingTransactions("validator-x");

        // Rewrite history: replace the transfer with a doubled amount, keep everything else.
        List<Block> tampered = new ArrayList<>(blockchain.blocks());
        Block original = tampered.get(1);
        Transaction tx = original.transactions().getFirst();
        Transaction doubled = new Transaction(tx.id(), tx.timestamp(), tx.sender(),
                tx.recipient(), new BigDecimal("6"), tx.type(), tx.payload(),
                tx.senderPublicKey(), tx.signature());
        tampered.set(1, new Block(original.index(), original.timestamp(), original.previousHash(),
                original.merkleRoot(), List.of(doubled, original.transactions().get(1)),
                original.validatorId(), original.consensusType(), original.proof(),
                original.nonce(), original.hash()));

        assertThat(blockchain.isValidChain(tampered)).isFalse();
    }

    @Test
    void replaceChainAdoptsOnlyLongerValidChains() {
        // Build a longer chain in a second ledger instance.
        ConsensusEngine engine = new ConsensusEngine(
                List.of(new ProofOfWorkStrategy(), new ProofOfStakeStrategy(), new BftStrategy()),
                "POS");
        Blockchain remote = new Blockchain(engine,
                new TokenomicsService(new BigDecimal("50"), 100, new BigDecimal("21000")),
                signatures, event -> {
        });
        remote.minePendingTransactions("remote-validator");
        remote.minePendingTransactions("remote-validator");

        assertThat(blockchain.replaceChain(remote.blocks())).isTrue();
        assertThat(blockchain.length()).isEqualTo(3);

        // A shorter chain must be refused.
        assertThat(blockchain.replaceChain(List.of(blockchain.blocks().getFirst()))).isFalse();
    }
}

File: app02_cpp_block/tests/BlockchainTest.cpp
// Proves the "real ledger, not a mock" properties: signed submission, sealing
// with rewards, full-chain audit, tamper detection and the longest-valid-chain
// sync rule. Mirrors app01_java_block's BlockchainTest.java.
#include <gtest/gtest.h>

#include <memory>
#include <vector>

#include "consensus/BftStrategy.h"
#include "consensus/ConsensusEngine.h"
#include "consensus/ProofOfStakeStrategy.h"
#include "consensus/ProofOfWorkStrategy.h"
#include "core/Blockchain.h"
#include "core/IdGen.h"
#include "crypto/PqcSignatureService.h"

using namespace legalchain;

namespace {
std::vector<std::unique_ptr<consensus::ConsensusStrategy>> allStrategies() {
    std::vector<std::unique_ptr<consensus::ConsensusStrategy>> v;
    v.push_back(std::make_unique<consensus::ProofOfWorkStrategy>());
    v.push_back(std::make_unique<consensus::ProofOfStakeStrategy>());
    v.push_back(std::make_unique<consensus::BftStrategy>());
    return v;
}
} // namespace

class BlockchainTest : public ::testing::Test {
protected:
    crypto::PqcSignatureService signatures;
    core::EventBus events;
    std::unique_ptr<consensus::ConsensusEngine> engine;
    std::unique_ptr<core::Blockchain> blockchain;
    crypto::SignatureKeyPair sender;
    std::string senderAddress;

    void SetUp() override {
        engine = std::make_unique<consensus::ConsensusEngine>(allStrategies(), "POS");
        core::TokenomicsService tokenomics(50.0, 100, 21000.0);
        blockchain = std::make_unique<core::Blockchain>(*engine, tokenomics, signatures, events);
        sender = signatures.generateKeyPair();
        senderAddress = crypto::PqcSignatureService::fingerprint(sender.publicKey);
    }

    core::Transaction signedTransfer(const std::string& recipient, double amount) {
        core::Transaction unsigned_(core::newUuid(), core::nowMillis(), senderAddress, recipient, amount,
                                     core::TransactionType::TRANSFER, {},
                                     crypto::PqcSignatureService::encodePublicKey(sender.publicKey), std::nullopt);
        core::Transaction signed_(unsigned_.id(), unsigned_.timestampMs(), unsigned_.sender(),
                                   unsigned_.recipient(), unsigned_.amount(), unsigned_.type(),
                                   unsigned_.payload(), unsigned_.senderPublicKey(),
                                   signatures.sign(unsigned_.contentToSign(), sender.secretKey));
        return signed_;
    }
};

TEST_F(BlockchainTest, StartsWithDeterministicGenesis) {
    EXPECT_EQ(blockchain->length(), 1);
    EXPECT_EQ(blockchain->blocks().front().previousHash(), std::string(64, '0'));
    EXPECT_TRUE(blockchain->isValid());
}

TEST_F(BlockchainTest, RejectsUnsignedAndForgedTransactions) {
    core::Transaction unsigned_(core::newUuid(), core::nowMillis(), senderAddress, "bob", 1.0,
                                 core::TransactionType::TRANSFER, {}, std::nullopt, std::nullopt);
    EXPECT_THROW(blockchain->submit(unsigned_), std::invalid_argument);

    // Valid signature but claimed sender is not the signing key's fingerprint.
    core::Transaction impersonation(core::newUuid(), core::nowMillis(), "someone-else", "bob", 1.0,
                                     core::TransactionType::TRANSFER, {},
                                     crypto::PqcSignatureService::encodePublicKey(sender.publicKey), std::nullopt);
    core::Transaction signedImpersonation(
        impersonation.id(), impersonation.timestampMs(), impersonation.sender(), impersonation.recipient(),
        impersonation.amount(), impersonation.type(), impersonation.payload(), impersonation.senderPublicKey(),
        signatures.sign(impersonation.contentToSign(), sender.secretKey));
    EXPECT_THROW(blockchain->submit(signedImpersonation), std::invalid_argument);
}

TEST_F(BlockchainTest, MinesBlockWithRewardAndDerivesBalances) {
    blockchain->submit(signedTransfer("bob", 3.0));
    core::Block block = blockchain->minePendingTransactions("validator-x");

    EXPECT_EQ(block.index(), 1);
    EXPECT_EQ(block.transactions().size(), 2u); // transfer + reward
    EXPECT_TRUE(blockchain->isValid());
    EXPECT_TRUE(blockchain->pendingTransactions().empty());

    auto balances = blockchain->balances();
    EXPECT_DOUBLE_EQ(balances["bob"], 3.0);
    EXPECT_DOUBLE_EQ(balances["validator-x"], 50.0);
}

TEST_F(BlockchainTest, DetectsTamperedHistory) {
    blockchain->submit(signedTransfer("bob", 3.0));
    blockchain->minePendingTransactions("validator-x");

    // Rewrite history: replace the transfer with a doubled amount, keep everything else.
    std::vector<core::Block> tampered = blockchain->blocks();
    core::Block original = tampered[1];
    core::Transaction tx = original.transactions().front();
    core::Transaction doubled(tx.id(), tx.timestampMs(), tx.sender(), tx.recipient(), 6.0, tx.type(), tx.payload(),
                               tx.senderPublicKey(), tx.signature());
    std::vector<core::Transaction> txs = {doubled, original.transactions()[1]};
    tampered[1] = core::Block(original.index(), original.timestampMs(), original.previousHash(),
                               original.merkleRoot(), txs, original.validatorId(), original.consensusType(),
                               original.proof(), original.nonce(), original.hash());

    EXPECT_FALSE(blockchain->isValidChain(tampered));
}

TEST_F(BlockchainTest, ReplaceChainAdoptsOnlyLongerValidChains) {
    // Build a longer chain in a second ledger instance.
    auto remoteEngine = std::make_unique<consensus::ConsensusEngine>(allStrategies(), "POS");
    core::TokenomicsService tokenomics(50.0, 100, 21000.0);
    core::EventBus remoteEvents;
    core::Blockchain remote(*remoteEngine, tokenomics, signatures, remoteEvents);
    remote.minePendingTransactions("remote-validator");
    remote.minePendingTransactions("remote-validator");

    EXPECT_TRUE(blockchain->replaceChain(remote.blocks()));
    EXPECT_EQ(blockchain->length(), 3);

    // A shorter chain must be refused.
    std::vector<core::Block> shortChain = {blockchain->blocks().front()};
    EXPECT_FALSE(blockchain->replaceChain(shortChain));
}
